geo/src/algorithm/triangulate_spade.rs

```diff
@@ -0,0 +1,568 @@
+impl<T: GeoFloat + SpadeNum> SpadeTriangulationFloat for T {}
+
+// ======= Triangulation trait =========
+
+pub type Triangles<T> = Vec<Triangle<T>>;
+
+/// Triangulate polygons using a [denlauny triangulation](https://en.wikipedia.org/wiki/Delaunay_triangulation)
+pub trait TriangulateSpade<'a, T>
+where
+    T: SpadeTriangulationFloat,
+{
+    /// collect all the lines that are relevant for triangulations from the geometric object that
+    /// should be triangulated.
+    ///
+    /// intersecting lines are allowed
+    fn get_lines(&'a self) -> Vec<Line<T>>;
+    /// collect all the points that are relevant for triangulations from the geometric object that
+    /// should be triangulated
+    fn get_points(&'a self) -> Vec<Coord<T>>;
+    /// define a predicate that decides if a point is inside of the object (used for constrained triangulation)
+    fn contains_point(&'a self, p: Point<T>) -> bool;
+
+    // processing of the lines that prepare the lines for triangulation.
+    //
+    // `spade` has the general
+    // limitation that constraint lines cannot intersect or else it will panic. This is why we need
+    // to manually split up the lines into smaller parts at the intersection point
+    //
+    // there's also a preprocessing step which tries to minimize the risk of failure of the algo
+    // through edge cases (thin/flat triangles are prevented as much as possible & lines are deduped, ...)
+    fn cleanup_lines(lines: Vec<Line<T>>) -> TriangulationResult<Vec<Line<T>>> {
+        let (known_points, lines) = preprocess_lines(lines);
+        prepare_intersection_contraint(lines, known_points)
+    }
+
+    /// returns a triangulation that's solely based on the points of the geometric object
+    ///
+    /// The triangulation is guaranteed to be Delaunay
+    ///
+    /// Note that the lines of the triangulation don't necessarily follow the lines of the input
+    /// geometry. If you wish to achive that take a look at the `constrained_triangulation` and the
+    /// `constrained_outer_triangulation` functions
+    fn unconstrained_triangulation(&'a self) -> TriangulationResult<Triangles<T>> {
+        let points = self.get_points();
+        points
+            .into_iter()
+            .map(to_spade_point)
+            .try_fold(DelaunayTriangulation::<Point2<T>>::new(), |mut tris, p| {
+                tris.insert(p).map_err(TriangulationError::SpadeError)?;
+                Ok(tris)
+            })
+            .map(triangulation_to_triangles)
+    }
+
+    /// returns triangulation that's based on the points of the geometric object and also
+    /// incorporates the lines of the input geometry
+    ///
+    /// The triangulation is not guaranteed to be Delaunay because of the constraint lines
+    ///
+    /// This outer triangulation also contains triangles that are not included in the input
+    /// geometry if it wasn't convex. Here's an example:
+    ///
+    /// ```ignore
```

Comment:
I think these diagram blocks should be marked as `text`, not `ignore`.

Reply:
(You'll see the difference if you run `cargo doc --open` locally.)

Reply:
Thanks for that suggestion. I'll apply it immediately 👍🏼